lib/Runtime/Library/JavascriptString.cpp

```diff
@@ -2135,26 +2135,57 @@ namespace Js
     {
         charcount_t count = pThis->GetLength();
 
-        BufferStringBuilder builder(count, pThis->type->GetScriptContext());
-        const wchar_t *inStr = pThis->GetString();
-        wchar_t *outStr = builder.DangerousGetWritableBuffer();
+        const wchar_t* inStr = pThis->GetString();
+        const wchar_t* i = inStr; 
+        const wchar_t* iLim = i + count;
+        
+        // first, try to see if the string needs recasing at all
+        while(i < iLim)
+        {
+            
+            // copy the current char into a temporary variable, then case it
+            wchar_t inChar = *i;
+            wchar_t outChar = *i;
+            CharUpperBuffW(&outChar, 1);
+            
+            // if the output is different from the input, the string needs recasing from here
+            if(outChar != inChar) { break; }
+            
+            // otherwise, we can keep going on
+            i++;
+            
+        }
+        
+        // if we reached the end of the string and didn't need recasing, we can return immediately
+        if(i == iLim)
+        { 
+            return pThis;
+        }
+        
+        // if not, we have to copy the first part of the string, then case the rest
+        charcount_t countToSkip = i - inStr;
+        charcount_t countToConvert = count - countToSkip; 
 
+        // start by copying the whole string in a new buffer
+        BufferStringBuilder builder(count, pThis->type->GetScriptContext());
+        wchar_t* outStr = builder.DangerousGetWritableBuffer();
         wchar_t* outStrLim = outStr + count;
-        wchar_t *o = outStr;
+        wchar_t* o = outStr;
 
         while (o < outStrLim)
         {
             *o++ = *inStr++;
         }
-
+        
+        // then recase it
         if(toCase == ToUpper)
         {
 #if DBG
             DWORD converted =
 #endif
-                CharUpperBuffW(outStr, count);
+                CharUpperBuffW(outStr + countToSkip, countToConvert);
```

Comment:
@dreamer-dead You're right, `CharUpperBuffW` doesn't convert the characters that would increase the length of the string. We previously had our own mapping logic, which also didn't handle those characters, but decided to [replace](https://github.com/Microsoft/ChakraCore/pull/253) it with the Windows API to improve our Unicode support since this API handles a superset of our old implementation.

It doesn't look like we had a GitHub issue tracking these special characters, so I've created [one](https://github.com/Microsoft/ChakraCore/issues/421).